File: 0207-course-schedule/0207-course-schedule.cpp
class Solution {
public:
    bool canFinish(int numCourses, vector<vector<int>>& prerequisites) {
        vector<vector<int>> adj(numCourses);
        vector<bool> vis1(numCourses, 0), vis2(numCourses, 0);
        for(auto a : prerequisites) adj[a[1]].push_back(a[0]);
        for(int i = 0; i<numCourses; i++) if(cycle(i, vis1, vis2, adj)) return false;
        return true;
    }
    bool cycle(int node, vector<bool> &vis1, vector<bool> &vis2, vector<vector<int>> &adj) {
        if(vis2[node]) return true;
        if(vis1[node]) return false;
        vis1[node] = true;
        vis2[node] = true;
        for(int i : adj[node]) if(cycle(i, vis1, vis2, adj)) return true;
        vis2[node] = false;
        return false;
    }
};

File: 0207-course-schedule/0207-course-schedule.py
class Solution:
    def canFinish(self, numCourses: int, prerequisites: List[List[int]]) -> bool:
        adj = {}
        for i in range(0, numCourses):
            adj[i] = []
        for i in prerequisites:
            adj[i[1]].append(i[0])
        vis1 = [False] * numCourses
        vis2 = [False] * numCourses
        def cycle(node):
            if vis2[node]:
                return True;
            if vis1[node]:
                return False
            vis1[node] = True
            vis2[node] = True
            for i in adj[node]:
                if cycle(i):
                    return True
            vis2[node] = False
            return False
        for i in range(0, numCourses):
            if cycle(i):
                return False
        return True
